### PR TITLE
Split client library code from ciao-cli

### PR DESCRIPTION
**Pull request by rbradford**
In preparation for producing a new command line client split the code from ciao-cli that is responsible for constructing HTTP requests, handling responses and parsing it's output into a new client library with calls for all currently used requests.

### CODE REVIEW

File: ciao-cli/client.go [hunk 1]
```diff
@@ -321,3 +322,446 @@ func (client *Client) DeleteEvents() error {
...
+
+	err = client.unmarshalHTTPResponse(resp, &IPs)
+	if err != nil {
+		return IPs, errors.Wrap(err, "Error parsing HTTP response")
+	}
+
+	return IPs, err
+}
+
+// UnmapExternalIP unmaps the givevne address from the instance
```
Comment (tmarcu) [inline ciao-cli/client.go]: s/givevne/given

File: ciao-cli/client.go [hunk 2]
```diff
@@ -321,3 +322,446 @@ func (client *Client) DeleteEvents() error {
...
+		return errors.Wrap(err, "Error getting external IP resource")
+	}
+
+	resp, err := client.sendHTTPRequest("POST", url, nil, body, ver)
+	if err != nil {
+		return errors.Wrap(err, "Error making HTTP request")
+	}
+
+	if resp.StatusCode != http.StatusNoContent {
+		return fmt.Errorf("External IP map failed: %s", resp.Status)
```
Comment (tmarcu) [inline ciao-cli/client.go]: Why don't we call errors.Wrapf() here for consistency?

File: ciao-cli/external_ips.go [hunk 3]
```diff
@@ -257,26 +173,13 @@ func (cmd *externalIPUnMapCommand) parseArgs(args []string) []string {
 
 func (cmd *externalIPUnMapCommand) run(args []string) error {
 	if cmd.address == "" {
-		errorf("Missing required -instance parameter")
+		errorf("Missing required -address parameter")
```
Comment (markdryan) [inline ciao-cli/external_ips.go]: Why don't we return the error here?

File: ciao-cli/external_ips.go [hunk 4]
```diff
@@ -802,36 +526,23 @@ func (cmd *poolRemoveCommand) run(args []string) error {
...
-	}
-
-	if url == "" {
-		fatalf("Address not present")
-	}
-
-	ver := api.PoolsV1
+		_, network, err := net.ParseCIDR(cmd.subnet)
+		if err != nil {
+			fatalf(err.Error())
```
Comment (markdryan) [inline ciao-cli/external_ips.go]: Again why are we calling fatalf here?

File: ciao-cli/image.go [hunk 5]
```diff
@@ -111,7 +79,7 @@ func (cmd *imageAddCommand) run(args []string) error {
 		return errors.New("Missing required -file parameter")
 	}
 
-	_, err := os.Stat(cmd.file)
+	f, err := os.Open(cmd.file)
 	if err != nil {
 		fatalf("Could not open %s [%s]\n", cmd.file, err)
```
Comment (markdryan) [inline ciao-cli/image.go]: Another fatalf here.

File: ciao-cli/image.go [hunk 6]
```diff
@@ -111,7 +79,7 @@ func (cmd *imageAddCommand) run(args []string) error {
 		return errors.New("Missing required -file parameter")
 	}
 
-	_, err := os.Stat(cmd.file)
+	f, err := os.Open(cmd.file)
```
Comment (markdryan) [inline ciao-cli/image.go]: Does f get closed anywhere?

File: ciao-cli/client.go [hunk 7]
```diff
@@ -286,47 +281,86 @@ func (client *Client) checkPrivilege() bool {
...
-// ListEvents retrieves the events for either all or the desired tenant
-func (client *Client) ListEvents(tenantID string) (types.CiaoEvents, error) {
-	var events types.CiaoEvents
-	var url string
+func (client *Client) getResource(url string, content string, query []queryValue, result interface{}) error {
+	resp, err := client.sendHTTPRequest("GET", url, query, nil, content)
+	if err != nil {
+		return errors.Wrapf(err, "Error making HTTP request to %s", url)
+	}
+	defer resp.Body.Close()
```
Comment (markdryan) [inline ciao-cli/client.go]: resp.Body is going to get closed twice, once when this function exits and once when unmarshalHTTPResponse exits.   This behaviour is undefined.

File: ciao-cli/client.go [hunk 8]
```diff
@@ -286,47 +281,86 @@ func (client *Client) checkPrivilege() bool {
...
+		return errors.Wrap(err, "Error marshalling JSON")
+	}
 
-	resp, err := client.sendHTTPRequest("DELETE", url, nil, nil, "")
+	resp, err := client.sendHTTPRequest("POST", url, nil, bytes.NewReader(b), content)
 	if err != nil {
-		return errors.Wrap(err, "Error making HTTP request")
+		return errors.Wrapf(err, "Error making HTTP request to %s", url)
 	}
 	defer resp.Body.Close()
```
Comment (markdryan) [inline ciao-cli/client.go]: Body may get closed twice.

File: ciao-cli/client.go [hunk 9]
```diff
@@ -315,6 +315,25 @@ func (client *Client) deleteResource(url string, content string) error {
...
+	b, err := json.Marshal(request)
+	if err != nil {
+		return errors.Wrap(err, "Error marshalling JSON")
+	}
+
+	resp, err := client.sendHTTPRequest("PUT", url, nil, bytes.NewReader(b), content)
+	if err != nil {
+		return errors.Wrapf(err, "Error making HTTP request to %s", url)
+	}
+	defer resp.Body.Close()
```
Comment (markdryan) [inline ciao-cli/client.go]: No need for the defer.

File: ciao-cli/main.go [hunk 10]
```diff
@@ -113,7 +114,7 @@ const (
 	ciaoClientCertFileEnv = "CIAO_CLIENT_CERT_FILE"
 )
 
-var client Client
+var c client.Client
```
Comment (markdryan) [inline ciao-cli/main.go]: Is c a good name for a global variable?

File: ciao-cli/client.go (hunk 9 above)
Comment (rbradford) [reply ciao-cli/client.go]: I decided to keep the defer here and removed it from the json unmarhall helper

File: ciao-cli/image.go (hunk 6 above)
Comment (rbradford) [reply ciao-cli/image.go]: Fixed

File: ciao-cli/client.go (hunk 7 above)
Comment (rbradford) [reply ciao-cli/client.go]: I didn't realise unmarshalHTTPResponse() did that. I think it's bad style to i've made sure that wherever a response is generated the code there is responsible for cleaning up.

**Pull request merged**